opds.py

```diff
@@ -1598,16 +1590,27 @@ def acquisition_link(cls, rel, href, types, active_loan=None):
         # and will not ask patrons to enter their passphrases
         # For more information please look here:
         # https://readium.org/lcp-specs/notes/lcp-key-retrieval.html#including-a-hashed-passphrase-in-an-opds-1-catalog
-        if active_loan and active_loan.license_pool.collection.protocol == ExternalIntegration.LCP:
+        if active_loan and active_loan.license_pool.collection.protocol in [
+            ExternalIntegration.LCP,
+            ExternalIntegration.ODL,
+            ExternalIntegration.ODL2,
+        ]:
             db = Session.object_session(active_loan)
             lcp_credential_factory = LCPCredentialFactory()
-            hashed_passphrase = lcp_credential_factory.get_hashed_passphrase(db, active_loan.patron)
 
-            hashed_passphrase_element = AtomFeed.makeelement(
-                "{%s}hashed_passphrase" % AtomFeed.LCP_NS)
-            hashed_passphrase_element.text = hashed_passphrase
+            try:
+                hashed_passphrase = lcp_credential_factory.get_hashed_passphrase(
+                    db, active_loan.patron
+                )
+                hashed_passphrase_element = AtomFeed.makeelement(
+                    "{%s}hashed_passphrase" % AtomFeed.LCP_NS
+                )
+                hashed_passphrase_element.text = hashed_passphrase
 
-            link.append(hashed_passphrase_element)
+                link.append(hashed_passphrase_element)
+            except LCPError:
+                # The patron's passphrase wasn't generated yet and not present in the database.
```

Comment:
Is this in expected case or should this never happen?

Reply:
This can happen easily. For example, when we deal with the collection with existing loans and fullfilments or it we deal with plain ODL collections. I don't like that we have to use exception handling instead of conditional logic but the inner implementation of `Credential` class and it `persistent_token_create` method  kind of forces us to do so.
I can refactor it but it'd require more changes in `LCPCredentialFactory` class.